Change references to CAMLP4 to CAMLP5 since we no longer use camlp4

jfehrle opened the pull request:
Infrastructure (cleanup)

I changed almost all the references to "camlp4" to "camlp5", did a "make clean", "./configure", "make" and "make byte" and ran coqtop successfully.  I'm waiting for the last Travis job to start/finish; the others succeeded.  I haven't run the makecoq_mingw.sh script locally (now Done).  Maybe @MSoegtropIMC can comment?

Notes:

- I didn't change camlp4 in comments/doc when it appeared to be historical info
- (Done) I didn't know the correct changes in a couple places in makecoq_mingw.sh so I didn't change them--please advise.  Lines 888-891
- (I think no) The file camlp4-4.02+6.patch should go away now?
- build-system.txt says

```
This classification replaces the old mechanism of declaring the use
of a grammar extension via a line of the form:
 (*i camlp5deps: "grammar.cma q_constr.cmo" i*)

The use of (*i camlp5use: ... i*) to mention uses of standard
extension such as IFDEF has also been discontinued, the Makefile now
always calls camlp5 with pa_macros.cmo and a few others by default.
```
(Yes, done) So perhaps I should remove those lines entirely from the changed .ml4 files?

- I still see the string "CAMLP4" in executable files coqchk.exe and coqmktop.exe.  How do I build these? 
 "make tools" doesn't rebuild them.  Should "make tools" do so?  (Also, how does a developer discover the "tools" target--should "make help" mention the "tools"?)

Zimmi48 commented on config/coq_config.mli:
```diff
@@ -28,11 +28,11 @@ val ocamllex : string
-val camlp4bin : string  (* base directory for Camlp4/5 binaries *)
-val camlp4lib : string  (* where is the library of Camlp4 *)
-val camlp4compat : string (* compatibility argument to camlp4/5 *)
+val camlp5 : string     (* exact name of camlp5: either "camlp5" ou "camlp5" *)
```
Some more refactoring could be done here.

Zimmi48 commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -928,7 +928,7 @@ function make_camlp5 {
   make_ocaml
   make_findlib
-  make_camlp4 # required by lablgtk-2.18.3 and lablgtk-2.18.5
+  make_camlp5 # required by lablgtk-2.18.3 and lablgtk-2.18.5
```
Watch out that the Windows build scripts do actually require camlp4 to build lablgtk.

Zimmi48 commented on dev/doc/build-system.txt:
```diff
@@ -123,11 +123,11 @@ There is now two kinds of preprocessed files :
 This classification replaces the old mechanism of declaring the use
 of a grammar extension via a line of the form:
- (*i camlp4deps: "grammar.cma q_constr.cmo" i*)
+ (*i camlp5deps: "grammar.cma q_constr.cmo" i*)
```
Keep this unchanged as this documented an old practice.

Zimmi48 commented on dev/doc/newsyntax.tex:
```diff
@@ -309,7 +309,7 @@ \subsection{Présentation de la grammaire}
 
 Les opérateurs infixes ou préfixes sont tous au même niveau de
-priorité du point de vue de Camlp4. La solution envisagée est de les
+priorité du point de vue de Camlp5. La solution envisagée est de les
```
Keep this file untouched as this is historical documentation.

Zimmi48 commented on plugins/btauto/g_btauto.ml4:
```diff
@@ -6,7 +6,7 @@
 (************************************************************************)
 
-(*i camlp4deps: "grammar/grammar.cma"  i*)
+(*i camlp5deps: "grammar/grammar.cma"  i*)
```
Just remove this line instead. It should work.

Zimmi48 commented on plugins/cc/g_congruence.ml4:
```diff
@@ -6,7 +6,7 @@
 (************************************************************************)
 
-(*i camlp4deps: "grammar/grammar.cma" i*)
+(*i camlp5deps: "grammar/grammar.cma" i*)
```
Just remove this line instead. It should work.

Zimmi48 commented on plugins/derive/g_derive.ml4:
```diff
@@ -8,7 +8,7 @@
 open Stdarg
 
-(*i camlp4deps: "grammar/grammar.cma"  i*)
+(*i camlp5deps: "grammar/grammar.cma"  i*)
```
Just remove this line instead. It should work.

Zimmi48 commented on plugins/extraction/g_extraction.ml4:
```diff
@@ -6,7 +6,7 @@
 (************************************************************************)
 
-(*i camlp4deps: "grammar/grammar.cma" i*)
+(*i camlp5deps: "grammar/grammar.cma" i*)
```
Etc.

Zimmi48 commented on plugins/ltac/g_obligations.ml4:
```diff
@@ -6,11 +6,11 @@
 (*
   Syntax for the subtac terms and types.
-  Elaborated from correctness/psyntax.ml4 by Jean-Christophe Filli�tre *)
+  Elaborated from correctness/psyntax.ml4 by Jean-Christophe Filli�tre *)
```
You seem to have an encoding issue. Probably due to Windows UTF-16. Keep to UTF-8 please.

jfehrle replied:
I'm not sure what is happening.
- My local git shows no change to this line
- https://github.com/jfehrle/coq/blob/master/plugins/ltac/g_obligations.ml4 shows the â rather than the �.
- When I examine the file with od, it contains the 3 byte sequence for �.  My editor shows the �, which I can replace with â, which then appears as a changed line in git.  That should fix it.

� is encoded as ef bf bd while â is c3 a2.  The former is actually in the file on my system:

```
0001160   s   t   o   p   h   e       F   i   l   l   i 357 277 275   t

0001160 73 74 6f 70 68 65 20 46 69 6c 6c 69 ef bf bd 74
```

FWIW: UTF-16 is mostly used in Asian countries with languages that have a very large character set such as Japanese and Chinese.  I've never encountered it.  Windows programs frequently use UTF-8 or one of the 8-bit ISO 8859 encodings.  This is clearly UTF-8, just this character is getting substituted somehow.

jfehrle replied:
Whoops, if I look in my branch cleanup_coq at https://github.com/jfehrle/coq/blob/master/plugins/ltac/g_obligations.ml4 it shows the �.

SkySkimmer replied:
emacs opens that file in iso-latin1 on my computer for master.

Zimmi48 commented on plugins/ssrmatching/ssrmatching.ml4:
```diff
@@ -12,8 +12,8 @@
-(*i camlp4use: "pa_extend.cmo" i*)
-(*i camlp4deps: "grammar/grammar.cma" i*)
+(*i camlp5use: "pa_extend.cmo" i*)
+(*i camlp5deps: "grammar/grammar.cma" i*)
```
Can someone confirm that these two lines can just be removed? Or only the second one?

MSoegtropIMC commented:
Hi Jim,

> Maybe @MSoegtropIMC<https://github.com/msoegtropimc> can comment?

what is documented in my build scrip (https://github.com/coq/coq/blob/master/dev/build/windows/makecoq_mingw.sh) is that lablgtk depends on camlp4. See function make_lablgtk and the notes on its dependency on make_camlp4.

As far as I remember, the configure script of lablgtk fails if camlp5 but not camlp4 is installed. Lablgtk is required for CoqIDE.

As far as I can tell from my scripts, this is the only dependency on camlp4.

Shall I try a more recent version of lablgtk without camlp4?

Best regards,

Michael
Intel Deutschland GmbH
Registered Address: Am Campeon 10-12, 85579 Neubiberg, Germany
Tel: +49 89 99 8853-0, www.intel.de
Managing Directors: Christin Eisenschmid, Christian Lamprechter
Chairperson of the Supervisory Board: Nicole Lau
Registered Office: Munich
Commercial Register: Amtsgericht Muenchen HRB 186928

jfehrle commented:
@MSoegtropIMC I'm trying to do only light cleanup.  Let's **not** switch to a newer version of labgtk as part of this PR.  Though that may be a good thing to try later.

Being slightly lazy, I hoped that one of the Travis builds would run makecoq_mingw.sh, so that if Travis is happy then the script is good.  Do you or @Zimmi48 know whether that is the case?  If not, I'll run it on my system in a clean directory.

jfehrle commented on dev/base_include:
```diff
@@ -19,7 +19,6 @@
 #directory "stm";;
 #directory "vernac";;
 
-#directory "+camlp4";; (* lazy solution: add both of camlp4/5 so that *)
```
Not sure I should remove this.

jfehrle commented:
BTW, developers will need to run ./configure again after pulling this change.  Maybe also "make clean".

MSoegtropIMC commented:
Dear Jim,

> Being slightly lazy, I hoped that one of the Travis builds would run makecoq_mingw.sh, so that if Travis is happy then the script is good.

> Do you or @Zimmi48<https://github.com/zimmi48> know whether that is the case? If not, I'll run it on my system in a clean directory.
I fixed the issues with Travis a while ago but didn’t track it – I was fairly busy in my job the last 6 months, hope it gets better soon. I will check the status the next days, latest over the weekend. I will also check if camlp4 is still required by recent versions of lablgtk.

Best regards,

Michael

Intel Deutschland GmbH
Registered Address: Am Campeon 10-12, 85579 Neubiberg, Germany
Tel: +49 89 99 8853-0, www.intel.de
Managing Directors: Christin Eisenschmid, Christian Lamprechter
Chairperson of the Supervisory Board: Nicole Lau
Registered Office: Munich
Commercial Register: Amtsgericht Muenchen HRB 186928

Zimmi48 commented:
Travis doesn't support Windows. That's why we also have AppVeyor. The packaging scripts are run there for 32bits and 64bits.

Zimmi48 commented:
The build here https://ci.appveyor.com/project/coq/coq/build/master~1946 had passed for 64bits but not for 32bits. Given that the failure looked weird and probably unrelated to this PR, I relaunched the build.

jfehrle commented:
I fixed the use of tab characters issue that Travis was complaining about.  Maybe good to go now?

FWIW, I see there are 18898 uses of the tab character in the code base (only looking at .ml and .mli files).  Do we want to eventually remove these--and how important is it to do so?  It's fairly easy for me to change these (1 command/file) but not sure I want to take this on now.

![image](https://user-images.githubusercontent.com/1253341/35589602-42e69656-05b9-11e8-8f99-8dee29788748.png)

SkySkimmer commented:
>FWIW, I see there are 18898 uses of the tab character in the code base (only looking at .ml and .mli files). Do we want to eventually remove these--and how important is it to do so? It's fairly easy for me to change these (1 command/file) but not sure I want to take this on now.

We considered removing them but didn't do it, see #6225.

jfehrle commented:
@Zimmi48 The failing job is failing while compiling a file that's outside Coq, so maybe this is now ready to merge?

MSoegtropIMC commented:
I somehow can't open the failing test on CircleCI (I tried a few times but it always times out) so I can't check if this is something which is in my domain (the Windows build scripts). I guess it is outside of my domain because afaik the Windows CI tests run on appveyor and the appveyor CI passes, but CircleCI is new to me and I have no idea what runs there and without being able to look at the logs I can't tell.

So if one of the builds using my cygwin based Windows build scripts (the variant not using opam) fail, please let me know and I will tyr my best to get logs from CircleCI and fix it.

Best regards,

Michael

SkySkimmer commented:
The failing build is equations. That's linux.
I had the same failure in  #6128 so it's probably a synchronization with master issue.

Zimmi48 commented:
The failing build is indeed expected because of the merge of https://github.com/mattam82/Coq-Equations/pull/46. To access the logs, one need to create a CircleCI account (for instance by logging in with GitHub). That's quite unfortunate.

Zimmi48 commented on config/coq_config.mli:
```diff
@@ -28,11 +28,11 @@ val ocamllex : string
+val camlp5o : string    (* name of the camlp5o/camlp5o executable *)
+val camlp5bin : string  (* base directory for Camlp5/5 binaries *)
+val camlp5lib : string  (* where is the library of Camlp5 *)
+val camlp5compat : string (* compatibility argument to camlp5/5 *)
```
These comments have slashes for camlp4/5 that could be removed.

Zimmi48 commented on config/coq_config.mli:
```diff
@@ -28,11 +28,11 @@ val ocamllex : string
-val camlp4bin : string  (* base directory for Camlp4/5 binaries *)
-val camlp4lib : string  (* where is the library of Camlp4 *)
-val camlp4compat : string (* compatibility argument to camlp4/5 *)
+val camlp5 : string     (* exact name of camlp5, i.e. "camlp5" *)
```
This variable and the logic associated to it can probably be removed since it only served to choose between camlp4 and camlp5.

ejgallego replied:
AFAICT this was not related to choice but to install location in pre-ocamlfind times.

Zimmi48 replied:
The two suggested values in the original code are "camlp5" and "camlp4".

ejgallego replied:
Oh indeed I misread the comment as affecting all the 3 lines, you are right.

ejgallego replied:
I get confused sometimes by the github context, I wish there was a way to mark the exact span of one is commenting.

jfehrle replied:
Would it make your life easier if I submitted these additional changes as a separate commit, with the idea that you could squash 2 commits into one before merging?  That way you'd only have to review the additional changes (and not re-read the earlier ones).

Zimmi48 replied:
Both solutions are fine to me.

Zimmi48 commented on configure.ml:
```diff
@@ -1095,11 +1095,11 @@ let write_configml f =
-  pr_s "camlp4bin" camlpXbindir;
-  pr_s "camlp4lib" camlpXlibdir;
-  pr_s "camlp4compat" camlp4compat;
+  pr_s "camlp5" camlpX;
```
All the variables named `camlpX` could probably be renamed into `camlp5` as well.

Zimmi48 commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -794,10 +794,10 @@ function make_ocaml {
+    # If we refer to the make variable ${PREFIX} below, camlp5 ends up having the wrong path:
     # D:\bin\coq64_buildtest_abs_ocaml4\bin>ocamlc -where => D:/bin/coq64_buildtest_abs_ocaml4/libocaml
-    # D:\bin\coq64_buildtest_abs_ocaml4\bin>camlp4 -where => ${PREFIX}/libocaml\camlp4
+    # D:\bin\coq64_buildtest_abs_ocaml4\bin>camlp5 -where => ${PREFIX}/libocaml\camlp5
```
I don't know whether the change `camlp4` -> `camlp5` in these comments is relevant or not.

jfehrle replied:
I will revert this change.  These look specific to a single developer's environment.  \"coq64_buildtest_abs_ocaml4" doesn't appear anywhere else in the code base.  Not every Windows system has a D: drive (\"D:\bin\...").  Also \"coq64_buildtest_abs_ocaml**4**\bin>camlp**5**" looks pretty odd.

Zimmi48 commented on dev/doc/build-system.txt:
```diff
@@ -125,9 +125,9 @@ This classification replaces the old mechanism of declaring the use
  (*i camlp4deps: "grammar.cma q_constr.cmo" i*)
 
-The use of (*i camlp4use: ... i*) to mention uses of standard
+The use of (*i camlp5use: ... i*) to mention uses of standard
```
Don't change this line since it's historical.

Zimmi48 commented on plugins/ssrmatching/ssrmatching.ml4:
```diff
@@ -12,8 +12,7 @@
 
-(*i camlp4use: "pa_extend.cmo" i*)
-(*i camlp4deps: "grammar/grammar.cma" i*)
+(*i camlp5use: "pa_extend.cmo" i*)
```
This directive was also discontinued according to `dev/doc/build-system.txt`, so just remove the line.

Zimmi48 commented on configure.ml:
```diff
@@ -623,23 +623,23 @@ let check_camlp5_version camlp5o =
-let camlpX, capitalized_camlpX, camlpXo,
-    camlpXbindir, fullcamlpXlibdir,
-    camlpXmod, camlpX_version = config_camlpX ()
+let capitalized_camlp5, camlp5o,
```
I would also (consistently) remove the `capitalized_camlp5`. I don't know what the story of this variable is but it looks like it serves no special purpose today and its constant value could be in-lined like you did for `camlp5`.

jfehrle commented:
@Zimmi48 Offhand, I don't see how the test failures could be related to my changes, maybe just random failures.  How can I rerun them?  Thanks.

Zimmi48 commented:
Indeed, the failing builds seem unrelated. In particular, the fiat-crypto issue can be observed as well in the test of the latest master branch: https://circleci.com/workflow-run/e5b4fb3e-ed7c-4287-a72a-96a17f0ed6e8 (that's worrying!)

jfehrle commented:
Yikes.  Git changes what should be trivial into something complex :-(.   All I wanted to do was pull in the latest changes from master to see if that would fix some of the test failures.  2 hours later, I don't understand why I see

```
Conflicting files
Makefile.build
lib/envars.ml
tools/CoqMakefile.in
```

or have any idea how I would fix that.

(Also still waiting for fiat-crypto to work again in main.)

SkySkimmer commented:
Why are file permissions changing?

jfehrle commented:
I'm not sure how the permissions changes crept in.  I fixed them to match master but I still see the same set of conflicting files:

```
$ git branch
* cleanup_camlp4
  master

Jim@DESKTOP-MMPNVGH /cygdrive/c/coq3/coq
$ ls -l Makefile.build lib/envars.ml lib/flags.ml lib/flags.mli tools/CoqMakefile.in tools/coqmktop.ml
-rw-rw-r--+ 1 Jim None  7214 Feb 11 16:39 lib/envars.ml
-rw-rw-r--+ 1 Jim None  6383 Feb 11 16:39 lib/flags.ml
-rw-rw-r--+ 1 Jim None  5992 Feb 11 16:39 lib/flags.mli
-rw-rw-r--+ 1 Jim None 27605 Feb 11 16:39 Makefile.build
-rw-r--r--+ 1 Jim None 25252 Feb 11 16:39 tools/CoqMakefile.in
-rw-rw-r--+ 1 Jim None 10930 Feb 11 16:39 tools/coqmktop.ml

Jim@DESKTOP-MMPNVGH /cygdrive/c/coq3/coq
$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

Jim@DESKTOP-MMPNVGH /cygdrive/c/coq3/coq
$ ls -l Makefile.build lib/envars.ml lib/flags.ml lib/flags.mli tools/CoqMakefile.in tools/coqmktop.ml
-rw-rw-r--+ 1 Jim None  7354 Feb 11 16:46 lib/envars.ml
-rw-rw-r--+ 1 Jim None  6383 Feb 11 16:46 lib/flags.ml
-rw-rw-r--+ 1 Jim None  5992 Feb 11 16:46 lib/flags.mli
-rw-rw-r--+ 1 Jim None 27605 Feb 11 16:46 Makefile.build
-rw-rw-r--+ 1 Jim None 25366 Feb 11 16:46 tools/CoqMakefile.in
-rw-rw-r--+ 1 Jim None 10932 Feb 11 16:46 tools/coqmktop.ml
```

jfehrle commented:
Oopa, missed this one, will fix:

```
-rw-r--r--+ 1 Jim None 25252 Feb 11 16:39 tools/CoqMakefile.in
```

jfehrle commented:
Also, it's odd that Travis info doesn't appear on this page even though it is indeed running builds for each push.

Zimmi48 commented:
I don't know what you did but you apparently rebased onto some older version of master. You now have many conflicts. I see that `Makefile.build` still refers to `API.mli` in your branch which shouldn't be the case anymore. Did you make sure that you pulled master from upstream (https://github.com/coq/coq) and not from your own fork (https://github.com/jfehrle/coq)? It could explain these problems.
I suggest you `git reset --hard f6fea244454ce3b8a0d4759fdbfa805e56315e8f` (state before your rebase) then properly pull the current master and rebase on top of it.

jfehrle commented:
Thanks, Théo.  Looks a lot better now.

The failing jobs were failing for the same reason in master.  Is there anything I should be doing with the branch, or will you or others pull PR when master is good again?

FWIW, I was pointing to the right branch with:
```
git remote add upstream https://github.com/coq/coq
```

But I should have done this before the pull:

```
git branch --set-upstream-to=upstream/master master
```

Zimmi48 commented:
> The failing jobs were failing for the same reason in master.

I'm not sure to see what you mean. AFAICT all jobs are green in master (Travis, AppVeyor and CircleCI).

jfehrle commented:
I was going off your comment from 6 days ago.  Actually I don't know how to see build jobs of master in coq/coq.

> In particular, the fiat-crypto issue can be observed as well in the test of the latest master branch: https://circleci.com/workflow-run/e5b4fb3e-ed7c-4287-a72a-96a17f0ed6e8 (that's worrying!)

Zimmi48 commented:
So in fact, I observed this issue only once in master, at about the same time as in this PR, so I guess it came from the fiat-crypto repository itself and has been fixed since then. To find the latest builds for master, click on the badges in the README.

jfehrle commented:
OMG!  This finally ran and passed all the tests, only taking 48 hours to complete!  Please merge, thanks.

Zimmi48 commented:
Actually CircleCI reports this job having waited for 93 hours until final completion. Yes, our CI is completely overloaded right now!

Zimmi48 commented on dev/doc/newsyntax.tex:
```diff
@@ -498,7 +498,7 @@ \subsubsection{Infixes extensibles}
 \subsubsection{Gestion des précédences}
 
-Les infixes peuvent être soit laissé à Camlp4, ou bien (comme ici)
+Les infixes peuvent être soit laissé à Camlp5, ou bien (comme ici)
```
I apparently missed this change in a file that is kept for historic reasons but shouldn't be changed.
Given how our CI is crowded, I guess if you update this commit to remove just this change, @maximedenes can agree to merge without waiting again for all the tests to pass...

jfehrle replied:
Done.
